### PR TITLE
[GBM] rework drm interface

### PR DESCRIPTION
**Pull request by lrusak**
This is the big drm refactor that I alluded to in https://github.com/xbmc/xbmc/pull/17710#issuecomment-616810518

This PR does quite a few things but maintains existing functionality:

 - move all drm components to the windowin/gbm/drm sub folder
 - move drm structs to classes with inheritance over the base class
 - reorganizes functions based on OOP to avoid having to call down many layers.
 - updates logging to the new style
 - some minor code cleanups

I didn't do doxygen yet as it will be a bit more intensive and this work might still change before that.

I have made it so that if any drm struct fails to be evaluated we hard fail.

There is still more code improvements to come but this should be enough for now

### CODE REVIEW

File: xbmc/windowing/gbm/drm/DRMCrtc.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+class CDRMCrtc : public CDRMObject
+{
+public:
+  explicit CDRMCrtc(int fd, uint32_t crtc);
+  CDRMCrtc(const CDRMCrtc&) = delete;
+  CDRMCrtc& operator=(const CDRMCrtc&) = delete;
+  ~CDRMCrtc() = default;
+
+  uint32_t& GetCrtcId() const { return m_crtc->crtc_id; }
```
Comment (garbear) [inline xbmc/windowing/gbm/drm/DRMCrtc.h:28]: It's a bit misleading to have a const function return a mutable result, as the function can be used in a const context, but the object can be mutated without const checks by assigning to the result reference, bypassing const-correctness.

For better const-correctness, the `const` here should be dropped. If const access is needed, then the idiomatic approach is to have two functions, one for each constness (one const and one mutable)

**Comment by garbear** [comment]
Code looks good. I like the abstractions you've built over the DRM functionality. I study the code and reflect that although readability is the in eye of the beholder, and this beholder reads C++ easier than C, I feel that readability objectively suffers when the C stuff is interfacing to the C++ stuff. There's arguments against verbosity, but having distinct layers, and having C code look like C and C++ code look like C++, would lead to beautiful code. This PR takes a step in that direction.

File: xbmc/windowing/gbm/drm/DRMCrtc.h (hunk 1 above)
Comment (a1rwulf) [reply xbmc/windowing/gbm/drm/DRMCrtc.h:28]: I never thought about it this way.
For my understanding the const just makes sure nobody accidently modifies the class state in such method.

Comment (garbear) [reply xbmc/windowing/gbm/drm/DRMCrtc.h:28]: How about this:

```c++
uint32_t GetCrtcId() const { return m_crtc->crtc_id; }
void SetCrtcId(uint32_t crtcId) { m_crtc->crtc_id = crtcId; }
```

Comment (lrusak) [reply xbmc/windowing/gbm/drm/DRMCrtc.h:28]: The setter is not needed. Basically I'm providing getters for the underlying C-struct that shouldn't be modified.

Comment (garbear) [reply xbmc/windowing/gbm/drm/DRMCrtc.h:28]: You shouldn't pass uint32_t by reference. the compiler can perform drastically superior optimizations of ints. I think what you want is:

```c++
uint32_t GetCrtcId() const { return m_crtc->crtc_id; }
```

Comment (a1rwulf) [reply xbmc/windowing/gbm/drm/DRMCrtc.h:28]: Yeah I see 2 possibilities:
`uint32_t GetCrtcId() const { return m_crtc->crtc_id; } // doesn't prevent developer from changing his crtc_id, but never changes the value in the class instance (you hand out a copy)`

Taken from your linked article:
`std::remove_reference_t<uint32_t> const uint32_t& GetCrtcId() const { return m_crtc->crtc_id; }`

Should give you an immutable reference, haven't tried though.

Comment (a1rwulf) [reply xbmc/windowing/gbm/drm/DRMCrtc.h]: imho just remove the ref and use plain int instead.

**Pull request merged**